dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/indicator/IndicatorTypeController.java

```diff
@@ -27,17 +27,53 @@
  */
 package org.hisp.dhis.webapi.controller.indicator;
 
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
 import org.hisp.dhis.common.OpenApi;
+import org.hisp.dhis.dxf2.webmessage.WebMessage;
+import org.hisp.dhis.dxf2.webmessage.WebMessageUtils;
+import org.hisp.dhis.feedback.MergeReport;
 import org.hisp.dhis.indicator.IndicatorType;
+import org.hisp.dhis.merge.MergeQuery;
+import org.hisp.dhis.merge.MergeQueryProcessor;
+import org.hisp.dhis.merge.MergeService;
+import org.hisp.dhis.merge.MergeType;
 import org.hisp.dhis.schema.descriptors.IndicatorTypeSchemaDescriptor;
 import org.hisp.dhis.webapi.controller.AbstractCrudController;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.access.prepost.PreAuthorize;
 import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
 @OpenApi.Tags("metadata")
 @Controller
 @RequestMapping(value = IndicatorTypeSchemaDescriptor.API_ENDPOINT)
-public class IndicatorTypeController extends AbstractCrudController<IndicatorType> {}
+@RequiredArgsConstructor
+@Slf4j
+public class IndicatorTypeController extends AbstractCrudController<IndicatorType> {
+
+  private final MergeService<IndicatorType> indicatorTypeMergeService;
+
+  @ResponseStatus(HttpStatus.OK)
+  @PreAuthorize("hasRole('ALL') or hasRole('F_INDICATOR_TYPE_MERGE')")
+  @PostMapping(value = "/merge", produces = APPLICATION_JSON_VALUE)
+  public @ResponseBody WebMessage mergeIndicatorTypes(@RequestBody MergeQuery query) {
+    log.info("Indicator type merge request received");
+
+    MergeReport report =
+        new MergeQueryProcessor<>(indicatorTypeMergeService)
+            .processMergeQuery(query, MergeType.INDICATOR_TYPE);
+
+    log.info("Indicator type merge request processed: {}", report);
+    return WebMessageUtils.mergeReport(report);
```

Comment:
I know this pattern of having a report handling method is used and established but it does hide possible error outcomes from OpenAPI. I think we should move towards having the `mergeReport` method throw a exception from the feedback package instead of directly generating different `WebMessage`s. Only the success message should be returned. The error cases can go via exceptions which in the end generate the same error web messages but in a way that is visible.

Reply:
update possible thrown exceptions to include merge report, if merge report has errors then throw conflict exception